squire_sdk/src/client/builder.rs

```diff
@@ -80,7 +77,13 @@ impl<UP: OnUpdate> ClientBuilder<UP, String, SquireAccount> {
             on_update,
         } = self;
         let client = Client::builder().build()?;
-        let resp = client.get(format!("{url}{VERSION_ROUTE}")).send().await?;
+        let resp = client
+            .get(format!(
+                "{url}{}",
+                <GetVersion as GetRequest<0>>::ROUTE.as_str()
```

Comment:
> Maybe this could be resolved by requiring `GetRequest: RestRequest`

That's a good idea. I did not try that. I'll do so. I do, however, fear that we will run into one of the aforementioned hurdles at the language level.

> This also feels like a shortcoming of the route API

I know that you are making a larger point here, but if we want to get rid of the explicit `as_str` call, we could implement `Display` for `Url` and have it simply display the held string.

Reply:
One last point here, we need the `GetRequest`, `PostRequest`, etc traits for the client. This is because we can't restrict the generic bounds of a function because off of associated constants. I looked high and low for a way to do something like this but I'm fairly certain that it is (certainly) impossible:
```rust
pub trait RestRequest<const N: usize>: Serialize + DeserializeOwned {
    const ROUTE: Url<N>;
    const METHOD: Method;
}

impl SquireClient {
    async fn get_request<const N: usize, B>(body: B, reps: [&str; N]) -> B::Response
        where B: RestRequest<N, METHOD = Method::Get> { ... }
}
```

Reply:
> we could implement `Display` for `Url` and have it simply display the held string.

Sorry, I was talking about two things at once. `Display` is already implemented for `Url` equivalent to `Url::as_str`, but the shortcoming I'm referring to is that `GetVersion` must be specified to be a `GetRequest`, and the route it could take on could be different from the route it takes on as a `PostRequest`.

Reply:
Ah, I understand what you're saying. Unfortunately, that is a shortcoming of this design. You could, in theory, implement GET and POST for the same type and/or implement GET twice of the same type (but with different const-generics).

While this is less than desirable, it is a step in the right direction. With this design, there is a single point of failure: implementing the request traits. As long as we avoid the mistakes that you describe, the trait system will help us avoid using the wrong routes. Compare this to how we were going things. Before, we just had a series of constant URLs that (ideally) had a name that matched their API types. This meant that for every type you defined an API or called one, you could make a mistake.